// bizcharts/lib/plots/core/dependents.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Scrollbar = exports.Slider = exports.Annotation = exports.Crosshair = exports.Component = exports.GroupComponent = exports.HtmlComponent = void 0;
var component_1 = require("@antv/component");
Object.defineProperty(exports, "Annotation", { enumerable: true, get: function () { return component_1.Annotation; } });
Object.defineProperty(exports, "Component", { enumerable: true, get: function () { return component_1.Component; } });
Object.defineProperty(exports, "Crosshair", { enumerable: true, get: function () { return component_1.Crosshair; } });
Object.defineProperty(exports, "GroupComponent", { enumerable: true, get: function () { return component_1.GroupComponent; } });
Object.defineProperty(exports, "HtmlComponent", { enumerable: true, get: function () { return component_1.HtmlComponent; } });
Object.defineProperty(exports, "Slider", { enumerable: true, get: function () { return component_1.Slider; } });
Object.defineProperty(exports, "Scrollbar", { enumerable: true, get: function () { return component_1.Scrollbar; } });
var g_base_1 = require("@antv/g-base");
Object.defineProperty(exports, "Event", { enumerable: true, get: function () { return g_base_1.Event; } });
Object.defineProperty(exports, "AbstractGroup", { enumerable: true, get: function () { return g_base_1.AbstractGroup; } });
Object.defineProperty(exports, "AbstractShape", { enumerable: true, get: function () { return g_base_1.AbstractShape; } });
var factory_1 = require("@antv/adjust/lib/factory");
Object.defineProperty(exports, "registerAdjust", { enumerable: true, get: function () { return factory_1.registerAdjust; } });
Object.defineProperty(exports, "getAdjust", { enumerable: true, get: function () { return factory_1.getAdjust; } });
Object.defineProperty(exports, "Adjust", { enumerable: true, get: function () { return factory_1.Adjust; } });
var factory_2 = require("@antv/attr/lib/factory");
Object.defineProperty(exports, "getAttribute", { enumerable: true, get: function () { return factory_2.getAttribute; } });
Object.defineProperty(exports, "Attribute", { enumerable: true, get: function () { return factory_2.Attribute; } });
var attr_1 = require("@antv/attr");
Object.defineProperty(exports, "Color", { enumerable: true, get: function () { return attr_1.Color; } });
var coord_1 = require("@antv/coord");
Object.defineProperty(exports, "getCoordinate", { enumerable: true, get: function () { return coord_1.getCoordinate; } });
Object.defineProperty(exports, "registerCoordinate", { enumerable: true, get: function () { return coord_1.registerCoordinate; } });
Object.defineProperty(exports, "Coordinate", { enumerable: true, get: function () { return coord_1.Coordinate; } });
var scale_1 = require("@antv/scale");
Object.defineProperty(exports, "getScale", { enumerable: true, get: function () { return scale_1.getScale; } });
Object.defineProperty(exports, "registerScale", { enumerable: true, get: function () { return scale_1.registerScale; } });
Object.defineProperty(exports, "Scale", { enumerable: true, get: function () { return scale_1.Scale; } });
