// bizcharts/lib/plots/plots.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// g2-plot 封装
var LineChart_1 = require("./LineChart");
Object.defineProperty(exports, "LineChart", { enumerable: true, get: function () { return LineChart_1.default; } });
var TreemapChart_1 = require("./TreemapChart");
Object.defineProperty(exports, "TreemapChart", { enumerable: true, get: function () { return TreemapChart_1.default; } });
var StepLineChart_1 = require("./StepLineChart");
Object.defineProperty(exports, "StepLineChart", { enumerable: true, get: function () { return StepLineChart_1.default; } });
var BarChart_1 = require("./BarChart");
Object.defineProperty(exports, "BarChart", { enumerable: true, get: function () { return BarChart_1.default; } });
var StackedBarChart_1 = require("./StackedBarChart");
Object.defineProperty(exports, "StackedBarChart", { enumerable: true, get: function () { return StackedBarChart_1.default; } });
var GroupedBarChart_1 = require("./GroupedBarChart");
Object.defineProperty(exports, "GroupedBarChart", { enumerable: true, get: function () { return GroupedBarChart_1.default; } });
var PercentStackedBarChart_1 = require("./PercentStackedBarChart");
Object.defineProperty(exports, "PercentStackedBarChart", { enumerable: true, get: function () { return PercentStackedBarChart_1.default; } });
var RangeBarChart_1 = require("./RangeBarChart");
Object.defineProperty(exports, "RangeBarChart", { enumerable: true, get: function () { return RangeBarChart_1.default; } });
var AreaChart_1 = require("./AreaChart");
Object.defineProperty(exports, "AreaChart", { enumerable: true, get: function () { return AreaChart_1.default; } });
var StackedAreaChart_1 = require("./StackedAreaChart");
Object.defineProperty(exports, "StackedAreaChart", { enumerable: true, get: function () { return StackedAreaChart_1.default; } });
var PercentStackedAreaChart_1 = require("./PercentStackedAreaChart");
Object.defineProperty(exports, "PercentStackedAreaChart", { enumerable: true, get: function () { return PercentStackedAreaChart_1.default; } });
var ColumnChart_1 = require("./ColumnChart");
Object.defineProperty(exports, "ColumnChart", { enumerable: true, get: function () { return ColumnChart_1.default; } });
var GroupedColumnChart_1 = require("./GroupedColumnChart");
Object.defineProperty(exports, "GroupedColumnChart", { enumerable: true, get: function () { return GroupedColumnChart_1.default; } });
var StackedColumnChart_1 = require("./StackedColumnChart");
Object.defineProperty(exports, "StackedColumnChart", { enumerable: true, get: function () { return StackedColumnChart_1.default; } });
var RangeColumnChart_1 = require("./RangeColumnChart");
Object.defineProperty(exports, "RangeColumnChart", { enumerable: true, get: function () { return RangeColumnChart_1.default; } });
var PercentStackedColumnChart_1 = require("./PercentStackedColumnChart");
Object.defineProperty(exports, "PercentStackedColumnChart", { enumerable: true, get: function () { return PercentStackedColumnChart_1.default; } });
var PieChart_1 = require("./PieChart");
Object.defineProperty(exports, "PieChart", { enumerable: true, get: function () { return PieChart_1.default; } });
var DensityHeatmapChart_1 = require("./DensityHeatmapChart");
Object.defineProperty(exports, "DensityHeatmapChart", { enumerable: true, get: function () { return DensityHeatmapChart_1.default; } });
var HeatmapChart_1 = require("./HeatmapChart");
Object.defineProperty(exports, "HeatmapChart", { enumerable: true, get: function () { return HeatmapChart_1.default; } });
var WordCloudChart_1 = require("./WordCloudChart");
Object.defineProperty(exports, "WordCloudChart", { enumerable: true, get: function () { return WordCloudChart_1.default; } });
var RoseChart_1 = require("./RoseChart");
Object.defineProperty(exports, "RoseChart", { enumerable: true, get: function () { return RoseChart_1.default; } });
var FunnelChart_1 = require("./FunnelChart");
Object.defineProperty(exports, "FunnelChart", { enumerable: true, get: function () { return FunnelChart_1.default; } });
var StackedRoseChart_1 = require("./StackedRoseChart");
Object.defineProperty(exports, "StackedRoseChart", { enumerable: true, get: function () { return StackedRoseChart_1.default; } });
var GroupedRoseChart_1 = require("./GroupedRoseChart");
Object.defineProperty(exports, "GroupedRoseChart", { enumerable: true, get: function () { return GroupedRoseChart_1.default; } });
var RadarChart_1 = require("./RadarChart");
Object.defineProperty(exports, "RadarChart", { enumerable: true, get: function () { return RadarChart_1.default; } });
var LiquidChart_1 = require("./LiquidChart");
Object.defineProperty(exports, "LiquidChart", { enumerable: true, get: function () { return LiquidChart_1.default; } });
var HistogramChart_1 = require("./HistogramChart");
Object.defineProperty(exports, "HistogramChart", { enumerable: true, get: function () { return HistogramChart_1.default; } });
var DonutChart_1 = require("./DonutChart");
Object.defineProperty(exports, "DonutChart", { enumerable: true, get: function () { return DonutChart_1.default; } });
var WaterfallChart_1 = require("./WaterfallChart");
Object.defineProperty(exports, "WaterfallChart", { enumerable: true, get: function () { return WaterfallChart_1.default; } });
var ScatterChart_1 = require("./ScatterChart");
Object.defineProperty(exports, "ScatterChart", { enumerable: true, get: function () { return ScatterChart_1.default; } });
var BubbleChart_1 = require("./BubbleChart");
Object.defineProperty(exports, "BubbleChart", { enumerable: true, get: function () { return BubbleChart_1.default; } });
var BulletChart_1 = require("./BulletChart");
Object.defineProperty(exports, "BulletChart", { enumerable: true, get: function () { return BulletChart_1.default; } });
var CalendarChart_1 = require("./CalendarChart");
Object.defineProperty(exports, "CalendarChart", { enumerable: true, get: function () { return CalendarChart_1.default; } });
var GaugeChart_1 = require("./GaugeChart");
Object.defineProperty(exports, "GaugeChart", { enumerable: true, get: function () { return GaugeChart_1.default; } });
var DualAxesChart_1 = require("./DualAxesChart");
Object.defineProperty(exports, "DualAxesChart", { enumerable: true, get: function () { return DualAxesChart_1.default; } });
