// @usedapp/core/dist/cjs/src/hooks/useSendTransaction.test.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
exports.__esModule = true;
var src_1 = require("../../src");
var chai_1 = require("chai");
var ethereum_waffle_1 = require("ethereum-waffle");
var ethers_1 = require("ethers");
var testing_1 = require("../../src/testing");
var BASE_TX_COST = 21000;
describe('useSendTransaction', function () {
    var mockProvider = new ethereum_waffle_1.MockProvider();
    var _a = mockProvider.getWallets(), spender = _a[0], receiver = _a[1], secondReceiver = _a[2];
    var network1;
    var config;
    var wallet1;
    var wallet2;
    beforeEach(function () { return __awaiter(void 0, void 0, void 0, function () {
        var _a;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0:
                    ;
                    return [4 /*yield*/, (0, testing_1.setupTestingConfig)()];
                case 1:
                    (_a = _b.sent(), config = _a.config, network1 = _a.network1);
                    wallet2 = network1.wallets[0];
                    wallet1 = ethers_1.ethers.Wallet.fromMnemonic('radar blur cabbage chef fix engine embark joy scheme fiction master release').connect(network1.provider);
                    return [4 /*yield*/, network1.wallets[1].sendTransaction({ to: wallet1.address, value: 100000 })];
                case 2:
                    _b.sent();
                    return [2 /*return*/];
            }
        });
    }); });
    it('success', function () { return __awaiter(void 0, void 0, void 0, function () {
        var _a, result, waitForCurrent, spenderBalance, receiverBalance, _b, _c;
        return __generator(this, function (_d) {
            switch (_d.label) {
                case 0: return [4 /*yield*/, (0, testing_1.renderWeb3Hook)(src_1.useSendTransaction, { mockProvider: mockProvider })];
                case 1:
                    _a = _d.sent(), result = _a.result, waitForCurrent = _a.waitForCurrent;
                    return [4 /*yield*/, spender.getBalance()];
                case 2:
                    spenderBalance = _d.sent();
                    return [4 /*yield*/, receiver.getBalance()];
                case 3:
                    receiverBalance = _d.sent();
                    return [4 /*yield*/, result.current.sendTransaction({ to: receiver.address, value: ethers_1.BigNumber.from(10), gasPrice: 0 })];
                case 4:
                    _d.sent();
                    return [4 /*yield*/, waitForCurrent(function (val) { return val.state !== undefined; })];
                case 5:
                    _d.sent();
                    (0, chai_1.expect)(result.current.state.status).to.eq('Success');
                    _b = chai_1.expect;
                    return [4 /*yield*/, receiver.getBalance()];
                case 6:
                    _b.apply(void 0, [_d.sent()]).to.eq(receiverBalance.add(10));
                    _c = chai_1.expect;
                    return [4 /*yield*/, spender.getBalance()];
                case 7:
                    _c.apply(void 0, [_d.sent()]).to.eq(spenderBalance.sub(10));
                    return [2 /*return*/];
            }
        });
    }); });
    it('sends with different signer', function () { return __awaiter(void 0, void 0, void 0, function () {
        var receiverBalance, secondReceiverBalance, _a, result, waitForCurrent, waitForNextUpdate, _b, _c;
        return __generator(this, function (_d) {
            switch (_d.label) {
                case 0: return [4 /*yield*/, receiver.getBalance()];
                case 1:
                    receiverBalance = _d.sent();
                    return [4 /*yield*/, secondReceiver.getBalance()];
                case 2:
                    secondReceiverBalance = _d.sent();
                    return [4 /*yield*/, (0, testing_1.renderWeb3Hook)(function () { return (0, src_1.useSendTransaction)({ signer: receiver }); }, {
                            mockProvider: mockProvider
                        })];
                case 3:
                    _a = _d.sent(), result = _a.result, waitForCurrent = _a.waitForCurrent, waitForNextUpdate = _a.waitForNextUpdate;
                    return [4 /*yield*/, waitForNextUpdate()];
                case 4:
                    _d.sent();
                    return [4 /*yield*/, result.current.sendTransaction({ to: secondReceiver.address, value: ethers_1.BigNumber.from(10) })];
                case 5:
                    _d.sent();
                    return [4 /*yield*/, waitForCurrent(function (val) { return val.state != undefined; })];
                case 6:
                    _d.sent();
                    (0, chai_1.expect)(result.current.state.status).to.eq('Success');
                    _b = chai_1.expect;
                    return [4 /*yield*/, secondReceiver.getBalance()];
                case 7:
                    _b.apply(void 0, [_d.sent()]).to.eq(secondReceiverBalance.add(10));
                    _c = chai_1.expect;
                    return [4 /*yield*/, receiver.getBalance()];
                case 8:
                    _c.apply(void 0, [_d.sent()]).to.not.eq(receiverBalance);
                    return [2 /*return*/];
            }
        });
    }); });
    it('Exception(invalid sender)', function () { return __awaiter(void 0, void 0, void 0, function () {
        var _a, result, waitForCurrent, waitForNextUpdate;
        return __generator(this, function (_b) {
            switch (_b.label) {
                case 0: return [4 /*yield*/, (0, testing_1.renderWeb3Hook)(src_1.useSendTransaction, { mockProvider: mockProvider })];
                case 1:
                    _a = _b.sent(), result = _a.result, waitForCurrent = _a.waitForCurrent, waitForNextUpdate = _a.waitForNextUpdate;
                    return [4 /*yield*/, waitForNextUpdate()];
                case 2:
                    _b.sent();
                    return [4 /*yield*/, result.current.sendTransaction({ to: '0x1', value: ethers_1.utils.parseEther('1') })];
                case 3:
                    _b.sent();
                    return [4 /*yield*/, waitForCurrent(function (val) { return val.state !== undefined; })];
                case 4:
                    _b.sent();
                    (0, chai_1.expect)(result.current.state.status).to.eq('Exception');
                    (0, chai_1.expect)(result.current.state.errorMessage).to.eq('invalid address');
                    return [2 /*return*/];
            }
        });
    }); });
    it('transfer ether with limit', function () { return __awaiter(void 0, void 0, void 0, function () {
        var _a, result, waitForCurrent, waitForNextUpdate;
        var _b;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0: return [4 /*yield*/, (0, testing_1.renderDAppHook)(function () { return (0, src_1.useSendTransaction)({ signer: wallet1 }); }, {
                        config: __assign(__assign({}, config), { gasLimitBufferPercentage: 100 })
                    })];
                case 1:
                    _a = _c.sent(), result = _a.result, waitForCurrent = _a.waitForCurrent, waitForNextUpdate = _a.waitForNextUpdate;
                    return [4 /*yield*/, waitForNextUpdate()];
                case 2:
                    _c.sent();
                    return [4 /*yield*/, result.current.sendTransaction({ to: wallet2.address, value: ethers_1.BigNumber.from(10), gasPrice: 0 })];
                case 3:
                    _c.sent();
                    return [4 /*yield*/, waitForCurrent(function (val) { return val.state !== undefined; })];
                case 4:
                    _c.sent();
                    (0, chai_1.expect)(result.current.state.status).to.eq('Success');
                    (0, chai_1.expect)((_b = result.current.state.transaction) === null || _b === void 0 ? void 0 : _b.gasLimit.toNumber()).to.equal(2 * BASE_TX_COST);
                    return [2 /*return*/];
            }
        });
    }); });
    it('transfer ether with limit in args', function () { return __awaiter(void 0, void 0, void 0, function () {
        var _a, result, waitForCurrent, waitForNextUpdate;
        var _b;
        return __generator(this, function (_c) {
            switch (_c.label) {
                case 0: return [4 /*yield*/, (0, testing_1.renderDAppHook)(function () {
                        return (0, src_1.useSendTransaction)({
                            signer: wallet1,
                            gasLimitBufferPercentage: 100
                        });
                    }, {
                        config: config
                    })];
                case 1:
                    _a = _c.sent(), result = _a.result, waitForCurrent = _a.waitForCurrent, waitForNextUpdate = _a.waitForNextUpdate;
                    return [4 /*yield*/, waitForNextUpdate()];
                case 2:
                    _c.sent();
                    return [4 /*yield*/, result.current.sendTransaction({ to: wallet2.address, value: ethers_1.BigNumber.from(10), gasPrice: 0 })];
                case 3:
                    _c.sent();
                    return [4 /*yield*/, waitForCurrent(function (val) { return val.state !== undefined; })];
                case 4:
                    _c.sent();
                    (0, chai_1.expect)(result.current.state.status).to.eq('Success');
                    (0, chai_1.expect)(result.current.state.status).to.eq('Success');
                    (0, chai_1.expect)((_b = result.current.state.transaction) === null || _b === void 0 ? void 0 : _b.gasLimit.toNumber()).to.equal(2 * BASE_TX_COST);
                    return [2 /*return*/];
            }
        });
    }); });
    it('Returns receipt after correct transaction', function () { return __awaiter(void 0, void 0, void 0, function () {
        var _a, result, waitForCurrent, spenderBalance, receiverBalance, _b, _c;
        var _d, _e, _f, _g, _h, _j, _k;
        return __generator(this, function (_l) {
            switch (_l.label) {
                case 0: return [4 /*yield*/, (0, testing_1.renderWeb3Hook)(src_1.useSendTransaction, { mockProvider: mockProvider })];
                case 1:
                    _a = _l.sent(), result = _a.result, waitForCurrent = _a.waitForCurrent;
                    return [4 /*yield*/, spender.getBalance()];
                case 2:
                    spenderBalance = _l.sent();
                    return [4 /*yield*/, receiver.getBalance()];
                case 3:
                    receiverBalance = _l.sent();
                    return [4 /*yield*/, result.current.sendTransaction({ to: receiver.address, value: ethers_1.BigNumber.from(10), gasPrice: 0 })];
                case 4:
                    _l.sent();
                    return [4 /*yield*/, waitForCurrent(function (val) { return val.state !== undefined; })];
                case 5:
                    _l.sent();
                    (0, chai_1.expect)(result.current.state.status).to.eq('Success');
                    _b = chai_1.expect;
                    return [4 /*yield*/, receiver.getBalance()];
                case 6:
                    _b.apply(void 0, [_l.sent()]).to.eq(receiverBalance.add(10));
                    _c = chai_1.expect;
                    return [4 /*yield*/, spender.getBalance()];
                case 7:
                    _c.apply(void 0, [_l.sent()]).to.eq(spenderBalance.sub(10));
                    (0, chai_1.expect)(result.current.state.receipt).to.not.be.undefined;
                    (0, chai_1.expect)((_d = result.current.state.receipt) === null || _d === void 0 ? void 0 : _d.to).to.eq(receiver.address);
                    (0, chai_1.expect)((_e = result.current.state.receipt) === null || _e === void 0 ? void 0 : _e.from).to.eq(spender.address);
                    (0, chai_1.expect)((_f = result.current.state.receipt) === null || _f === void 0 ? void 0 : _f.gasUsed).to.be.gt(0);
                    (0, chai_1.expect)((_g = result.current.state.receipt) === null || _g === void 0 ? void 0 : _g.status).to.eq(1);
                    (0, chai_1.expect)((_h = result.current.state.receipt) === null || _h === void 0 ? void 0 : _h.blockHash).to.match(/^0x/);
                    (0, chai_1.expect)((_j = result.current.state.receipt) === null || _j === void 0 ? void 0 : _j.transactionHash).to.match(/^0x/);
                    (0, chai_1.expect)((_k = result.current.state.receipt) === null || _k === void 0 ? void 0 : _k.gasUsed).to.be.gt(0);
                    return [2 /*return*/];
            }
        });
    }); });
    it('Can send transaction with private key', function () { return __awaiter(void 0, void 0, void 0, function () {
        var _a, result, waitForCurrent, waitForNextUpdate, spenderBalance, receiverBalance, _b, _c;
        var _d, _e, _f, _g, _h, _j, _k;
        return __generator(this, function (_l) {
            switch (_l.label) {
                case 0: return [4 /*yield*/, (0, testing_1.renderDAppHook)(function () { return (0, src_1.useSendTransaction)({ chainId: 1, privateKey: wallet1.privateKey }); }, { config: config })];
                case 1:
                    _a = _l.sent(), result = _a.result, waitForCurrent = _a.waitForCurrent, waitForNextUpdate = _a.waitForNextUpdate;
                    return [4 /*yield*/, waitForNextUpdate()];
                case 2:
                    _l.sent();
                    return [4 /*yield*/, wallet1.getBalance()];
                case 3:
                    spenderBalance = _l.sent();
                    return [4 /*yield*/, wallet2.getBalance()];
                case 4:
                    receiverBalance = _l.sent();
                    return [4 /*yield*/, result.current.sendTransaction({ to: wallet2.address, value: ethers_1.BigNumber.from(10), gasPrice: 0 })];
                case 5:
                    _l.sent();
                    return [4 /*yield*/, waitForCurrent(function (val) { return val.state !== undefined; })];
                case 6:
                    _l.sent();
                    (0, chai_1.expect)(result.current.state.status).to.eq('Success');
                    _b = chai_1.expect;
                    return [4 /*yield*/, wallet2.getBalance()];
                case 7:
                    _b.apply(void 0, [_l.sent()]).to.eq(receiverBalance.add(10));
                    _c = chai_1.expect;
                    return [4 /*yield*/, wallet1.getBalance()];
                case 8:
                    _c.apply(void 0, [_l.sent()]).to.eq(spenderBalance.sub(10));
                    (0, chai_1.expect)(result.current.state.receipt).to.not.be.undefined;
                    (0, chai_1.expect)((_d = result.current.state.receipt) === null || _d === void 0 ? void 0 : _d.to).to.eq(wallet2.address);
                    (0, chai_1.expect)((_e = result.current.state.receipt) === null || _e === void 0 ? void 0 : _e.from).to.eq(wallet1.address);
                    (0, chai_1.expect)((_f = result.current.state.receipt) === null || _f === void 0 ? void 0 : _f.gasUsed).to.be.gt(0);
                    (0, chai_1.expect)((_g = result.current.state.receipt) === null || _g === void 0 ? void 0 : _g.status).to.eq(1);
                    (0, chai_1.expect)((_h = result.current.state.receipt) === null || _h === void 0 ? void 0 : _h.blockHash).to.match(/^0x/);
                    (0, chai_1.expect)((_j = result.current.state.receipt) === null || _j === void 0 ? void 0 : _j.transactionHash).to.match(/^0x/);
                    (0, chai_1.expect)((_k = result.current.state.receipt) === null || _k === void 0 ? void 0 : _k.gasUsed).to.be.gt(0);
                    return [2 /*return*/];
            }
        });
    }); });
    it('Can send transaction with mnemonic phrase', function () { return __awaiter(void 0, void 0, void 0, function () {
        var _a, result, waitForCurrent, waitForNextUpdate, spenderBalance, receiverBalance, _b, _c;
        var _d, _e, _f, _g, _h, _j, _k;
        return __generator(this, function (_l) {
            switch (_l.label) {
                case 0: return [4 /*yield*/, (0, testing_1.renderDAppHook)(function () { return (0, src_1.useSendTransaction)({ chainId: 1, mnemonicPhrase: wallet1.mnemonic.phrase }); }, { config: config })];
                case 1:
                    _a = _l.sent(), result = _a.result, waitForCurrent = _a.waitForCurrent, waitForNextUpdate = _a.waitForNextUpdate;
                    return [4 /*yield*/, waitForNextUpdate()];
                case 2:
                    _l.sent();
                    return [4 /*yield*/, wallet1.getBalance()];
                case 3:
                    spenderBalance = _l.sent();
                    return [4 /*yield*/, wallet2.getBalance()];
                case 4:
                    receiverBalance = _l.sent();
                    return [4 /*yield*/, result.current.sendTransaction({ to: wallet2.address, value: ethers_1.BigNumber.from(10), gasPrice: 0 })];
                case 5:
                    _l.sent();
                    return [4 /*yield*/, waitForCurrent(function (val) { return val.state !== undefined; })];
                case 6:
                    _l.sent();
                    (0, chai_1.expect)(result.current.state.status).to.eq('Success');
                    _b = chai_1.expect;
                    return [4 /*yield*/, wallet2.getBalance()];
                case 7:
                    _b.apply(void 0, [_l.sent()]).to.eq(receiverBalance.add(10));
                    _c = chai_1.expect;
                    return [4 /*yield*/, wallet1.getBalance()];
                case 8:
                    _c.apply(void 0, [_l.sent()]).to.eq(spenderBalance.sub(10));
                    (0, chai_1.expect)(result.current.state.receipt).to.not.be.undefined;
                    (0, chai_1.expect)((_d = result.current.state.receipt) === null || _d === void 0 ? void 0 : _d.to).to.eq(wallet2.address);
                    (0, chai_1.expect)((_e = result.current.state.receipt) === null || _e === void 0 ? void 0 : _e.from).to.eq(wallet1.address);
                    (0, chai_1.expect)((_f = result.current.state.receipt) === null || _f === void 0 ? void 0 : _f.gasUsed).to.be.gt(0);
                    (0, chai_1.expect)((_g = result.current.state.receipt) === null || _g === void 0 ? void 0 : _g.status).to.eq(1);
                    (0, chai_1.expect)((_h = result.current.state.receipt) === null || _h === void 0 ? void 0 : _h.blockHash).to.match(/^0x/);
                    (0, chai_1.expect)((_j = result.current.state.receipt) === null || _j === void 0 ? void 0 : _j.transactionHash).to.match(/^0x/);
                    (0, chai_1.expect)((_k = result.current.state.receipt) === null || _k === void 0 ? void 0 : _k.gasUsed).to.be.gt(0);
                    return [2 /*return*/];
            }
        });
    }); });
    it('Can send transaction with encrypted json', function () { return __awaiter(void 0, void 0, void 0, function () {
        var json, _a, result, waitForCurrent, waitForNextUpdate, spenderBalance, receiverBalance, _b, _c;
        var _d, _e, _f, _g, _h, _j, _k;
        return __generator(this, function (_l) {
            switch (_l.label) {
                case 0: return [4 /*yield*/, wallet1.encrypt('test')];
                case 1:
                    json = _l.sent();
                    return [4 /*yield*/, (0, testing_1.renderDAppHook)(function () {
                            return (0, src_1.useSendTransaction)({
                                chainId: 1,
                                password: 'test',
                                json: json
                            });
                        }, { config: config })];
                case 2:
                    _a = _l.sent(), result = _a.result, waitForCurrent = _a.waitForCurrent, waitForNextUpdate = _a.waitForNextUpdate;
                    return [4 /*yield*/, waitForNextUpdate()];
                case 3:
                    _l.sent();
                    return [4 /*yield*/, wallet1.getBalance()];
                case 4:
                    spenderBalance = _l.sent();
                    return [4 /*yield*/, wallet2.getBalance()];
                case 5:
                    receiverBalance = _l.sent();
                    return [4 /*yield*/, result.current.sendTransaction({ to: wallet2.address, value: ethers_1.BigNumber.from(10), gasPrice: 0 })];
                case 6:
                    _l.sent();
                    return [4 /*yield*/, waitForCurrent(function (val) { return val.state !== undefined; })];
                case 7:
                    _l.sent();
                    (0, chai_1.expect)(result.current.state.status).to.eq('Success');
                    _b = chai_1.expect;
                    return [4 /*yield*/, wallet2.getBalance()];
                case 8:
                    _b.apply(void 0, [_l.sent()]).to.eq(receiverBalance.add(10));
                    _c = chai_1.expect;
                    return [4 /*yield*/, wallet1.getBalance()];
                case 9:
                    _c.apply(void 0, [_l.sent()]).to.eq(spenderBalance.sub(10));
                    (0, chai_1.expect)(result.current.state.receipt).to.not.be.undefined;
                    (0, chai_1.expect)((_d = result.current.state.receipt) === null || _d === void 0 ? void 0 : _d.to).to.eq(wallet2.address);
                    (0, chai_1.expect)((_e = result.current.state.receipt) === null || _e === void 0 ? void 0 : _e.from).to.eq(wallet1.address);
                    (0, chai_1.expect)((_f = result.current.state.receipt) === null || _f === void 0 ? void 0 : _f.gasUsed).to.be.gt(0);
                    (0, chai_1.expect)((_g = result.current.state.receipt) === null || _g === void 0 ? void 0 : _g.status).to.eq(1);
                    (0, chai_1.expect)((_h = result.current.state.receipt) === null || _h === void 0 ? void 0 : _h.blockHash).to.match(/^0x/);
                    (0, chai_1.expect)((_j = result.current.state.receipt) === null || _j === void 0 ? void 0 : _j.transactionHash).to.match(/^0x/);
                    (0, chai_1.expect)((_k = result.current.state.receipt) === null || _k === void 0 ? void 0 : _k.gasUsed).to.be.gt(0);
                    return [2 /*return*/];
            }
        });
    }); });
});
