// bizcharts/lib/g-components/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.render = void 0;
var reconciler_1 = __importDefault(require("./reconciler"));
var Canvas_1 = require("./Canvas");
Object.defineProperty(exports, "Canvas", { enumerable: true, get: function () { return Canvas_1.default; } });
var Group_1 = require("./Group");
Object.defineProperty(exports, "Group", { enumerable: true, get: function () { return Group_1.default; } });
var Circle_1 = require("./Circle");
Object.defineProperty(exports, "Circle", { enumerable: true, get: function () { return Circle_1.default; } });
var Ellipse_1 = require("./Ellipse");
Object.defineProperty(exports, "Ellipse", { enumerable: true, get: function () { return Ellipse_1.default; } });
var Image_1 = require("./Image");
Object.defineProperty(exports, "Image", { enumerable: true, get: function () { return Image_1.default; } });
var Line_1 = require("./Line");
Object.defineProperty(exports, "Line", { enumerable: true, get: function () { return Line_1.default; } });
var Marker_1 = require("./Marker");
Object.defineProperty(exports, "Marker", { enumerable: true, get: function () { return Marker_1.default; } });
var Path_1 = require("./Path");
Object.defineProperty(exports, "Path", { enumerable: true, get: function () { return Path_1.default; } });
var Polygon_1 = require("./Polygon");
Object.defineProperty(exports, "Polygon", { enumerable: true, get: function () { return Polygon_1.default; } });
var Polyline_1 = require("./Polyline");
Object.defineProperty(exports, "Polyline", { enumerable: true, get: function () { return Polyline_1.default; } });
var Rect_1 = require("./Rect");
Object.defineProperty(exports, "Rect", { enumerable: true, get: function () { return Rect_1.default; } });
var Text_1 = require("./Text");
Object.defineProperty(exports, "Text", { enumerable: true, get: function () { return Text_1.default; } });
exports.render = reconciler_1.default.render;
