// bizcharts/lib/components/Annotation/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
// annotation G2暂不支持自定义组件。
var base_1 = require("./base");
Object.defineProperty(exports, "Base", { enumerable: true, get: function () { return base_1.default; } });
var arc_1 = require("./arc");
Object.defineProperty(exports, "Arc", { enumerable: true, get: function () { return arc_1.default; } });
var dataMarker_1 = require("./dataMarker");
Object.defineProperty(exports, "DataMarker", { enumerable: true, get: function () { return dataMarker_1.default; } });
var dataRegion_1 = require("./dataRegion");
Object.defineProperty(exports, "DataRegion", { enumerable: true, get: function () { return dataRegion_1.default; } });
var regionFilter_1 = require("./regionFilter");
Object.defineProperty(exports, "RegionFilter", { enumerable: true, get: function () { return regionFilter_1.default; } });
var html_1 = require("./html");
Object.defineProperty(exports, "Html", { enumerable: true, get: function () { return html_1.default; } });
var reactElement_1 = require("./reactElement");
Object.defineProperty(exports, "ReactElement", { enumerable: true, get: function () { return reactElement_1.default; } });
var image_1 = require("./image");
Object.defineProperty(exports, "Image", { enumerable: true, get: function () { return image_1.default; } });
var line_1 = require("./line");
Object.defineProperty(exports, "Line", { enumerable: true, get: function () { return line_1.default; } });
var region_1 = require("./region");
Object.defineProperty(exports, "Region", { enumerable: true, get: function () { return region_1.default; } });
var text_1 = require("./text");
Object.defineProperty(exports, "Text", { enumerable: true, get: function () { return text_1.default; } });
